### ISSUE TITLE
Dry run flag

### ISSUE THREAD

**Issue by jackheizer** [rizsotto/Bear#284]
I'm not sure if this makes sense to request, but I was wondering if it would be possible to add a --dry-run flag to bear.

Essentially in some contexts (such as the clangd indexer) I don't necessarily want to compile the resulting binary to get the compilation database as that can be time consuming. Rather I'm wondering if it's possible for bear to intercept the calls to the compiler but not actually run them in order to construct the compilation database.

Thanks!

**Comment by rizsotto** [comment]
Hi @jackheizer , that's a valid use case. It's in the backlog for a while. The upcoming 3.x version might support it. (That version will have compiler wrappers, which would be able to do this. But it's still many unknowns in the story.)

One challenge is that the wrapper still need to produce output. Otherwise the build system (eg.: make) will not continue. Then need to wrap the linker and other tools too, to deal with the fake output of the wrappers. And these still might fail, if one target in the build is executed (eg.: to generate source file). So the technique has limitations, and requires more efforts now.

**Comment by Petross404** [comment]
Compiledb is another tool to generate compilation database and has this ability with `-n` flag. One can take a look [here](https://github.com/nickdiego/compiledb/blob/master/README.md).

**Comment by z-aki** [comment]
`xcodebuild` which is one use case for a tool like bear, supports `-dry-run` and it runs almost instantly for a build that otherwise takes 40 minutes.

**Comment by gianmarcoodorizzi** [comment]
> Compiledb is another tool to generate compilation database and has this ability with `-n` flag. One can take a look [here](https://github.com/nickdiego/compiledb/blob/master/README.md).

You're right. Too bad that that tools seems discontinued. Is there a workaround for this while we wait for an official implementation, @rizsotto?

**Comment by rizsotto** [comment]
@gianmarcoodorizzi the best workaround is to run the build. :)

**Comment by jbmaillet** [comment]
> One challenge is that the wrapper still need to produce output. Otherwise the build system (eg.: make) will not continue.

Not really linked to the build time problem, but If that can be useful to anyone: with good old GNU make, there are 2 interesting options that I use:

--ignore-errors Ignore all errors in commands executed to remake files.

--keep-going Continue as much as possible after an error.  While the target that failed, and those that depend on it, cannot be re-made, the other dependencies of these targets can be processed all the same.

In my use cases, I'm only interested in a compilation database for the open source components. These more or less always build OK, so it does not hurt. It's always _our_ proprietary code that is broken, and I don't need a compilation db for it for now (this may change in the future).

Of course this does not shorten the build time, or not by much, but at least I always get a usable output for my use case.

**Comment by max0x7ba** [comment]
This request for `bear` enhancement was closed as completed without a reference to more information.

Could you kindly refer me to documentation of `bear` command line options to enable and make use of this new enhancement functionality, please?

**Comment by rizsotto** [comment]
Good point. I've closed old tickets to track the new version progress. I might be implement this next week for the next release.

**Comment by afm0901x** [comment]
Hi, would it be possible to support piped stdin as an alternative to wrapper approach. I'll admit that I have no idea how `bear` currently intercept the build system call to compiler, but for this method atleast it would be pattern matching.

This would allow `make --dry-run --always-make ... | bear --output=...` to generate a compilation database without compiling the whole project. Additionally, any build system would also benefit from these as long as they can print out compilation command, e.g.:
```
# build.sh
#!/bin/sh
dry_run=
while getopts :d opt; do
  case $opt in
  d) dry_run=echo;;
  ?) exit 2;;
  esac
done
shift $((OPTIND-1))
$dry_run gcc -o app.exe main.c

# gen compilation db
$ ./build.sh -d | bear

# build
$ ./build.sh
```

The downside is it'll have to trust the build system to support dry-run and print out the command.

**Comment by rizsotto** [comment]
@jackheizer - thanks for raising this, and for everyone's patience on the
long thread.

I've implemented it as a new subcommand, `bear parse-sh`. Rather than a
`--dry-run` flag on interception - Bear watches real `exec()` calls, so it
can't observe a build that never runs - it reads the text your build already
prints in dry-run mode and feeds it to the normal analysis:

```sh
make -n | bear parse-sh | bear semantic --input -
```

- `make -n` prints the commands your build would run, without running them;
- `bear parse-sh` turns each of those lines into an execution event;
- `bear semantic --input -` reads the events and writes `compile_commands.json`.

It's best-effort - the database is only as complete as the dry-run text - so
`bear -- make` stays the recommended path when you can actually run the build.

It's on the `4.2.0-rc` branch, not in a release yet - I'd like some feedback
before I cut one. If you (and @Petross404, @z-aki, @gianmarcoodorizzi,
@afm0901x, @max0x7ba) could build that branch and try it on your projects,
that would help a lot.

**Comment by max0x7ba** [comment]
> It's on the `4.2.0-rc` branch, not in a release yet - I'd like some feedback before I cut one.

I added `TAGS2` target with your new commands into https://github.com/max0x7ba/atomic_queue

[`TAGS2`](https://github.com/max0x7ba/atomic_queue/blob/097be9f623dad373df4a3553092a386ca8a08f63/Makefile#L305) uses a locally built `bear` 4.2.0-rc installed into `~/.local/bin/bear`. The original [`TAGS`](https://github.com/max0x7ba/atomic_queue/blob/097be9f623dad373df4a3553092a386ca8a08f63/Makefile#L297) target uses your recommended `bear -- make` method.

Your new commands work as expected with no issues -- `TAGS` and `TAGS2` targets produce identical `compile_commands.json` files.

You should be able to reproduce my observations easily with:

```
git clone git@github.com:max0x7ba/atomic_queue.git
make -RC atomic_queue TAGS2
```

The original `TAGS` target has to build, and that requires cloning the 3rd-party dependencies for the build to succeed. 

Whereas `TAGS2` doesn't build and always succeeds -- and this is exactly what I inquired about your new `bear` functionality for. Thank you for implementing that 🙏🫡

**Comment by afm0901x** [comment]
@rizsotto - Thank you for your work on such useful program.

I tested the `4.2.0-rc` branch and it work fantastically for both `make -n` and the example I mentioned above.

The only quirky part is having to chain subcommand through shell and my suggestion would be:
```
$ bear                     # neither pipe nor args - print usage
$ bear <subcommand>        # only args - run subcommand
$ bear -- <build_command>  # end marker used - generate compilation db
$ make -n | bear           # pipe used - internally(cli frontend) run parse-sh+semantic to generate compilation db
```

**Comment by rizsotto** [comment]
Thanks @afm0901x for your feedback!

I am not sure if I want to implement all your suggestions. But I hear what you are saying about make the usage friction-less with less pipe and less flags.

**Comment by max0x7ba** [comment]
> The only quirky part is having to chain subcommand through shell and my suggestion would be:
> 
> ```
> $ make -n | bear           # pipe used - internally(cli frontend) run parse-sh+semantic to generate compilation db
> ```

What is the reason for your aversion to shell and pipes, may I ask?

Would not a simple `bash` function solve your woes? E.g.:

```
$ function bear2 { bear parse-sh | bear semantic --input - ; }
$ make -Bn | bear2
```

**Comment by rizsotto** [comment]
I consider to provide a single sub-command to produce the compilation database.

- `make -n | bear parse-sh` would produce the `compile_commands.json`.
- `bear` without sub-command already associated with the intercept+semantic mode. Adding the parse+semantic would be very different semantic.
